### PR TITLE
DDP flag to remove null gradients

### PR DESCRIPTION
**Pull request by borguz**
Summary: Distributed training hangs if some parameters don't receive gradients at every step.  Diff implements a flag to remove null gradients as a workaround.

Differential Revision: D13798663

### CODE REVIEW

File: torch/nn/parallel/distributed.py [hunk 1]
```diff
@@ -171,6 +171,8 @@ class DistributedDataParallel(Module):
                          are getting different gradients, which should not
                          happen if DistributedDataParallel is corrected used.
                          (default: False)
+        remove_null_grads: set to True to support cases where some paramters may
```
Comment (pietern) [inline torch/nn/parallel/distributed.py]: Typo: paramters

**Review by pietern — COMMENTED** [review]
Can you add a test for this as well?

**Review by apaszke — CHANGES_REQUESTED** [review]
Unfortunately this is a big can of worms, and we should be careful. This is only a partial implementation that allows us to omit a subset of parameters provided that *all processes will have the same set of unused parameters at every step*. Otherwise you'll get a deadlock/error/who knows. We should either improve the implementation or at least have a big warning sign in the docs.

**Comment by pietern** [comment]
@borguz I think this one can be abandoned, right? I recall you mentioned you found a workaround for this. Was it related to zero'ing gradients and letting them all participate?

**Comment by borguz** [comment]
Already closed internally, did not realize it's still open in github.

**Pull request closed without merging**